k-distribution/include/kframework/cparser/main.c

```diff
@@ -1,37 +1,102 @@
+#include "node.h"
+#include "parser.tab.h"
+#include "scanner.h"
+
+#include <assert.h>
+#include <stdarg.h>
 #include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include "node.h"
-#include "parser.tab.h"
-#include "scanner.h"
+
+struct string_buffer {
+  FILE *fp;
+  char *buf;
+  size_t capacity;
+  size_t idx;
+};
+
+struct string_buffer string_buffer_new_memory(size_t initial) {
+  assert(initial > 0 && "Invalid initial capacity");
+
+  char *buf = malloc(sizeof(*buf) * initial);
+  buf[0] = '\0';
+
+  struct string_buffer ret
+      = {.fp = NULL, .buf = buf, .capacity = initial, .idx = 0};
+  return ret;
+}
+
+struct string_buffer string_buffer_new_file(FILE *fp) {
+  struct string_buffer ret = {.fp = fp, .buf = NULL, .capacity = 0, .idx = 0};
+  return ret;
+}
+
+void string_buffer_grow(struct string_buffer *sb) {
+  if (sb->fp) {
+    return;
+  }
+
+  size_t old_cap = sb->capacity;
+  size_t new_cap = old_cap * 2;
+
+  sb->buf = realloc(sb->buf, new_cap);
+  sb->buf[old_cap] = 0;
+  sb->capacity = new_cap;
+}
+
+int buf_printf(struct string_buffer *sb, char const *format, ...) {
+  va_list args;
+
+  if (sb->fp) {
+    va_start(args, format);
+    int result = vfprintf(sb->fp, format, args);
+    va_end(args);
+    return result;
+  } else {
+    va_start(args, format);
+    int required = vsnprintf(NULL, 0, format, args) + 1;
+    va_end(args);
+
+    while ((sb->capacity - sb->idx) < required) {
+      string_buffer_grow(sb);
+    }
+
+    va_start(args, format);
```

Comment:
What is going on with this code? We should not be calling va_start or va_end multiple times in a single function call. There is way to copy va_lists if you need to use it more than once.

Reply:
Addressed